### DISCUSSION TITLE
neovim copy to clipboard kitty

### DISCUSSION THREAD

**Discussion by gaurishhs** [kovidgoyal/kitty#6640] [Q&A] [answered]
hi, i am unable to copy text from neovim editor to my clipboard. i even have a mapping setup in kitty.conf for `copy_to_clipboard` this is my current `kitty.conf` 

```
font_family      JetBrains Mono Regular
bold_font        JetBrains Mono Bold
italic_font      JetBrains Mono Italic
bold_italic_font JetBrains Mono Bold Italic

symbol_map U+23FB-U+23FE,U+2665,U+26A1,U+2B58,U+E000-U+E00A,U+E0A0-U+E0A3,U+E0B0-U+E0D4,U+E200-U+E2A9,U+E300-U+E3E3,U+E5FA-U+E6AA,U+E700-U+E7C5,U+EA60-U+EBEB,U+F000-U+F2E0,U+F300-U+F32F,U+F400-U+F4A9,U+F500-U+F8FF,U+F0001-U+F1AF0 Symbols Nerd Font Mono
map cmd+c copy_to_clipboard
```
however pressing `y` works. is there a way to map cmd+c to send y to neovim?

**Comment by kovidgoyal** [comment, marked as answer]
https://sw.kovidgoyal.net/kitty/faq/#how-do-i-map-key-presses-in-kitty-to-different-keys-in-the-terminal-program

Though if you want cmd+c to copy in neovim map it to yank in neovim.

**Reply by gaurishhs** [reply]
the thing is `vnoremap <D-c> "+y` isn't actually working for me even though ctrl+c works but cmd isn't working. surprisingly this thing does work in neovim gui (neovide). i also have clipboard=unnamed,unnamedplus set

**Reply by gaurishhs** [reply]
@kovidgoyal Any idea?

**Reply by kovidgoyal** [reply]
You need to ask for help in a neovim forum.

**Comment by GeorgeFlerovsky** [comment]
@gaurishhs @kovidgoyal 
I was able to make the Neovim clipboard work in kitty with the following settings (using OSC 52).
For more info, see `:help clipboard-osc52` in Neovim.
```lua
vim.opt.clipboard = ""
vim.g.clipboard = {
  name = 'OSC 52',
  copy = {
    ['+'] = require('vim.ui.clipboard.osc52').copy('+'),
    -- ['*'] = require('vim.ui.clipboard.osc52').copy('*'),
  },
  paste = {
    ['+'] = require('vim.ui.clipboard.osc52').paste('+'),
    -- ['*'] = require('vim.ui.clipboard.osc52').paste('*'),
  },
}
```

Note that due to Kitty's prompt for manual confirmation whenever pasting from the system clipboard, I don't recommend setting `vim.opt.clipboard` to "unnamed" or "unnamedplus"; otherwise, then you'll get the prompt whenever you copy and paste purely within Neovim. Instead, you should use the `"+y` and `"+p` key combos to copy and paste with the system clipboard.

I recommend adding this to kitty docs somewhere, so people don't have to ask about it in issues/discussion/reddit. I'm using:

- Wayland on Ubuntu 24.04.1 LTS
- Kitty 0.37.0
- Neovim version 0.10.2.

**Reply by TamaMcGlinn** [reply]
@GeorgeFlerovsky I really wish that worked. Unfortunately, Neovim just responds with "Waiting for OSC 52 response from the terminal" until it times out. I have also tried the snippet provided in `:help clipboard-osc52`, which has the same result. Help would be much appreciated, if anyone finds out how to do this.

**Reply by GeorgeFlerovsky** [reply]
@TamaMcGlinn 
Are you launching neovim from kitty terminal?

Also, do you have tmux running? Maybe it's better to do it without tmux


Which version of Neovim?

Are you on Wayland or X?

**Reply by TamaMcGlinn** [reply]
Thanks for the response!
Yes I'm just running kitty 0.35.2 with neovim 0.11.0 directly in that.
I'm on Ubuntu 22.04 LTS with X11

**Reply by GeorgeFlerovsky** [reply]
If you're on X, you don't need this OSC trick. I'm only using this trick because I'm on Wayland, which doesn't support the standard X way of hooking up the clipboard.

Instead, you should install `xsel` and/or `xclip`, which are linux clipboard providers for CLI programs in X11. Neovim will detect these if they are visible on your `$PATH` and will then be able to interact with the system clipboard.

- https://ostechnix.com/access-clipboard-contents-using-xclip-and-xsel-in-linux/

Excerpt from `:help clipboard-tool`:
```
							      *clipboard-tool*
The presence of a working clipboard tool implicitly enables the '+' and "*"
registers. Nvim looks for these clipboard tools, in order of priority:

  - |g:clipboard| (unless unset or `false`)
  - pbcopy, pbpaste (macOS)
  - wl-copy, wl-paste (if $WAYLAND_DISPLAY is set)
  - waycopy, waypaste (if $WAYLAND_DISPLAY is set)
  - xsel (if $DISPLAY is set)
  - xclip (if $DISPLAY is set)
  - lemonade (for SSH) https://github.com/pocke/lemonade
  - doitclient (for SSH) https://www.chiark.greenend.org.uk/~sgtatham/doit/
  - win32yank (Windows)
  - termux (via termux-clipboard-set, termux-clipboard-set)
  - tmux (if $TMUX is set)
```

**Reply by GeorgeFlerovsky** [reply]
Also, you'll need to use `"+y` to copy from Vim into the system clipboard and `"+p` to paste from the system clipboard.

Normally, most people set the `vim.opt.clipboard = "unnamedplus"` option in Neovim to make the usual `y` and `p` use the system clipboard directly, but this is not advised on kitty, because kitty requires manual user confirmation every time a CLI program tries to paste from the system clipboard, which would be quite annoying if you're just copying and pasting within Neovim (i.e. not to other programs). So I **don't** recommend setting it if you're running Neovim in kitty.

**Reply by TamaMcGlinn** [reply]
looks like you're right, all I had to do was remove the snippet I had gotten from the help message; I have `xclip 0.13-2` installed already, and it works again - also in tmux. I guess I was trying to figure out forwarding over SSH of the clipboard, found and tried out this solution to a different problem, and then left the snippet in place in my vimrc to confuse future me.

Thanks for your help, this has been bugging me for a long time!

**Reply by GeorgeFlerovsky** [reply]
@TamaMcGlinn 
I think Lemonade might be the tool you want for clipboard over ssh. I haven't used it myself, but it might work for you. Also, apparently Neovim can detect and use it automatically.

https://github.com/lemonade-command/lemonade

**Reply by kovidgoyal** [reply]
@GeorgeFlerovsky there's nothing kitty specific there it will work with any terminal that supports OSC 52. Note that you can turn off kitty's security prompt in kitty.conf via the clipboard_control setting if you like.

**Reply by GeorgeFlerovsky** [reply]
@kovidgoyal Agreed. I was just helping him make his setup work.

Personally, I prefer kitty's clipboard prompt. It's more secure. 💪
